### PR TITLE
Fix/add filesystem watcher event to queue

### PR DESCRIPTION
**Pull request by amanPras**
Bring back https://github.com/eclipse-che4z/che-che4z-lsp-for-cobol/pull/2443

### CODE REVIEW

File: clients/cobol-lsp-vscode-extension/src/test/suite/lsp.spec.test.ts [hunk 1]
```diff
@@ -151,7 +155,7 @@ suite("Integration Test Suite", function () {
       diagnostics[0].range,
       range(pos(58, 21), pos(58, 28)),
     );
-    assert.ok(diagnostics[0].message.includes("Syntax error on 'XCTL123'"));
+    assert.ok(diagnostics[0].message.includes("Extraneous input 'XCTL123'"));
```
Comment (ap891843) [inline clients/cobol-lsp-vscode-extension/src/test/suite/lsp.spec.test.ts:158]: I believe, this is some recent change

File: clients/cobol-lsp-vscode-extension/src/test/suite/lsp.spec.test.ts [hunk 2]
```diff
@@ -401,16 +399,15 @@ suite("Integration Test Suite", function () {
...
       editor,
       pos(18, 0),
       "           FD  TRANS-FILE-IN IS EXTERNAL.\n",
     );
     await helper.waitFor(
-      () => vscode.languages.getDiagnostics(editor.document.uri).length === 2,
+      () => vscode.languages.getDiagnostics(editor.document.uri).length > 0,
     );
     let diagnostics = vscode.languages.getDiagnostics(editor.document.uri);
     assert.strictEqual(
```
Comment (slavek-kucera) [inline clients/cobol-lsp-vscode-extension/src/test/suite/lsp.spec.test.ts:420]: Can you be 100% sure that the message you are waiting for will be the first entry in the array? Should the code search all the diags and try to find it?

File: clients/cobol-lsp-vscode-extension/src/test/suite/testHelper.ts [hunk 3]
```diff
@@ -78,31 +78,30 @@ export async function showDocument(workspace_file: string) {
...
+    "workbench.action.revertAndCloseActiveEditor",
+  );
 }
 
 export async function closeAllEditors() {
-  await waitFor(async () => {
-    if (vscode.window.activeTextEditor === undefined) {
-      return true;
-    }
+  let doc = vscode.window.activeTextEditor;
```
Comment (slavek-kucera) [inline clients/cobol-lsp-vscode-extension/src/test/suite/testHelper.ts]:
```suggestion
  let editor = vscode.window.activeTextEditor;
```
or maybe just 
```
while(vscode.window.activeTextEditor?.document)
```

File: server/common/src/main/java/org/eclipse/lsp/cobol/common/benchmark/BenchmarkServiceImpl.java [hunk 4]
```diff
@@ -24,12 +24,13 @@
...
 import java.nio.file.StandardOpenOption;
 import java.util.*;
+import java.util.concurrent.CopyOnWriteArrayList;
 import java.util.stream.Collectors;
 
 public class BenchmarkServiceImpl implements BenchmarkService {
   private final static Logger LOG = LoggerFactory.getLogger(BenchmarkServiceImpl.class);
   private static final String PERFORMANCE_LOG_PATH = "performance.log.path";
-  private final List<BenchmarkSession> benchmarkSessions = new ArrayList<>();
+  private final List<BenchmarkSession> benchmarkSessions = new CopyOnWriteArrayList<>();
```
Comment (slavek-kucera) [inline server/common/src/main/java/org/eclipse/lsp/cobol/common/benchmark/BenchmarkServiceImpl.java:33]: What code accesses this varible concurrently?

File: server/engine/src/main/java/org/eclipse/lsp/cobol/lsp/analysis/AsyncAnalysisService.java [hunk 5]
```diff
@@ -114,19 +114,36 @@ public synchronized CompletableFuture<CobolDocumentModel> scheduleAnalysis(Strin
...
-    CompletableFuture<CobolDocumentModel> value = CompletableFuture.supplyAsync(() -> {
+    ExecutorService analysisExecutor = getExecutor(uri);
+    CobolDocumentModel documentModel = documentModelService.get(uri);
+    documentModel.updateLastAnalysisNotCancelled();
+    FutureTask<CobolDocumentModel> futureTask = new FutureTask<>(scheduleAnalysis(uri, documentModel, text, currentRevision, open, force, eventSource, id));
+    analysisResults.put(id, futureTask);
+    analysisExecutor.submit(futureTask);
+    if (prevId != null && !force) {
+      Optional.ofNullable(analysisResults.get(makeId(uri, prevId))).ifPresent(cf -> cf.cancel(true));
+    }
```
Comment (slavek-kucera) [inline server/engine/src/main/java/org/eclipse/lsp/cobol/lsp/analysis/AsyncAnalysisService.java]: Shouldn't this be in a similar loop that can be found in `reanalyseOpenedPrograms`?

File: server/common/src/main/java/org/eclipse/lsp/cobol/common/benchmark/BenchmarkServiceImpl.java (hunk 4 above)
Comment (amanPras) [reply server/common/src/main/java/org/eclipse/lsp/cobol/common/benchmark/BenchmarkServiceImpl.java:33]: https://github.com/eclipse-che4z/che-che4z-lsp-for-cobol/blob/295158b35558a5aa30c26bbf4d9a91e141bc06a7/server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/CobolLanguageEngine.java#L165

This service is invoked by each analysis thread. Multiple thread could try to logtime

File: server/engine/src/main/java/org/eclipse/lsp/cobol/lsp/analysis/AsyncAnalysisService.java (hunk 5 above)
Comment (amanPras) [reply server/engine/src/main/java/org/eclipse/lsp/cobol/lsp/analysis/AsyncAnalysisService.java]: I am under impression that this method is responsible to scheduling an analysis and the caller of this method should try to cancel previous analysis before calling this method.
But I think, `reanalyseCopybooksAssociatedPrograms` should be updated

Comment (slavek-kucera) [reply server/engine/src/main/java/org/eclipse/lsp/cobol/lsp/analysis/AsyncAnalysisService.java]: Fine with me, but be also aware of the other direct calls to the `scheduleAnalysis` function - from didOpen, didChange handlers and AnalysisHandler.getDependencies.

Comment (slavek-kucera) [reply server/engine/src/main/java/org/eclipse/lsp/cobol/lsp/analysis/AsyncAnalysisService.java]: to be handled in #2467.

File: clients/cobol-lsp-vscode-extension/src/test/suite/lsp.spec.test.ts (hunk 2 above)
Comment (amanPras) [reply clients/cobol-lsp-vscode-extension/src/test/suite/lsp.spec.test.ts:420]: updated

File: clients/cobol-lsp-vscode-extension/src/test/suite/testHelper.ts (hunk 3 above)
Comment (amanPras) [reply clients/cobol-lsp-vscode-extension/src/test/suite/testHelper.ts]: updated

File: server/engine/src/main/java/org/eclipse/lsp/cobol/lsp/analysis/AsyncAnalysisService.java [hunk 6]
```diff
@@ -114,47 +116,71 @@ public synchronized CompletableFuture<CobolDocumentModel> scheduleAnalysis(Strin
...
+                        + analysisResultsRevisions.get(uri));
         return null;
       }
-      LOG.debug("[scheduleAnalysis] Start analysis: " + uri);
-      notifyAllListeners(AnalysisState.STARTED, documentModelService.get(uri), eventSource);
+      LOG.debug("[scheduleAnalysis] waiting for previous analysis of {} to finish", uri);
       try {
+        if (documentModel.getLastAnalysisResult() != null) {
+          waitForPreviousAnalysisToBeCompleted(ImmutableList.of(uri));
+        }
```
Comment (slavek-kucera) [inline server/engine/src/main/java/org/eclipse/lsp/cobol/lsp/analysis/AsyncAnalysisService.java]: Should the waiting be unconditional? The rule is that we don't want to cancel the first parsing, however we still don't want run two parsings side by side, right?

Comment (amanPras) [reply server/engine/src/main/java/org/eclipse/lsp/cobol/lsp/analysis/AsyncAnalysisService.java]: Yes, totally missed this scenario.

Updated in the latest commit

**Review by slavek-kucera — APPROVED** [review]

**Comment by amanPras** [comment]
waiting for manual test

File: server/engine/src/main/java/org/eclipse/lsp/cobol/lsp/analysis/AsyncAnalysisService.java [hunk 7]
```diff
@@ -154,21 +154,22 @@ private Callable<CobolDocumentModel> scheduleAnalysis(String uri, CobolDocumentM
       LOG.debug("[scheduleAnalysis] waiting for previous analysis of {} to finish", uri);
       try {
         waitForPreviousAnalysisToBeCompleted(ImmutableList.of(uri));
-        documentModel.markAnalysisInProgress();
+        documentModelService.get(uri).markAnalysisInProgress();
```
Comment (slavek-kucera) [inline server/engine/src/main/java/org/eclipse/lsp/cobol/lsp/analysis/AsyncAnalysisService.java]: Given the multithreading nature of this code, anytime you call the `documentModelService.get(uri)` you might get a different result, right?

Comment (amanPras) [reply server/engine/src/main/java/org/eclipse/lsp/cobol/lsp/analysis/AsyncAnalysisService.java]: Yes, I will revert this

Comment (amanPras) [reply server/engine/src/main/java/org/eclipse/lsp/cobol/lsp/analysis/AsyncAnalysisService.java]: updated

**Review by slavek-kucera — APPROVED** [review]

**Pull request merged**